batch/batch/front_end/front_end.py

```diff
@@ -956,7 +956,10 @@ async def _create_job_group(
 async def _create_job_groups(db: Database, batch_id: int, update_id: int, user: str, job_group_specs: List[dict]):
     assert len(job_group_specs) > 0
 
-    validate_job_groups(job_group_specs)
+    try:
+        validate_job_groups(job_group_specs)
+    except ValidationError as e:
+        raise web.HTTPBadRequest(reason=e.reason)
```

Comment:
Could this be moved up the stack? I feel like once execution reaches this point, the request itself should have been well-formed at least?

Reply:
Ya it could be moved up one level to the route handler. I changed the other resource creation methods to match. Slightly repetitive, but ensures we never touch the database if anything about the request is malformed. Unfortunately we're not using strict mode for pyright and `json_request` returns `Any` (so passing its output to other methods will always pass the typechecker regardless of the signature) so I'm not so sure how we can enforce in the type checker that the specs were validated.

Reply:
That's a nice improvement!